tests/jax/test_custom_partitioning.py

```diff
@@ -0,0 +1,301 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+import jax
+from jax import random
+import jax.numpy as jnp
+import numpy as np
+import pytest
+from functools import partial
+from jax.experimental.pjit import pjit
+
+from utils import is_devices_enough
+from flax.linen import make_attention_mask
+from flax.linen import make_causal_mask
+from flax.linen import combine_masks
+from transformer_engine.jax.flax import extend_logical_axis_rules
+from transformer_engine.jax.sharding import get_dot_sharding_meta
+from transformer_engine.jax.sharding import get_elementwise_sharding_meta
+from transformer_engine.jax.sharding import get_fp8_meta_sharding_meta
+from transformer_engine.jax.sharding import global_shard_guard
+from transformer_engine.jax.sharding import infer_major_sharding_type
+from transformer_engine.jax.sharding import is_dp_enabled, is_tp_enabled
+from transformer_engine.jax.sharding import ShardingMeta, ShardingResource, ShardingType
```

Comment:
`ShardingMeta`, `ShardingResource` and `ShardingType` are deprecating. Will fully removed after the migration to CP is done.

Reply:
I think we should keep the old map-based sharding around for a while, encapsulated into its own submodule/file and selectively loaded instead of the new CP-based sharding if we detect an older Jax version that don't support the new custom ops. We can deprecate and remove it permanently down the line, assuming Jax doesn't change its mind again on sharding.